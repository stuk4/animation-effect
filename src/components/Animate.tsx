import React, { cloneElement, useEffect } from 'react'
import { Children } from 'react';
import {  IAnimateProps } from '../interfaces/IAnimations';
import { addDelayCSSAnimation, addNewClassNamesAnimation } from '../utils/animations';


export const Animate = ({children,delay=0,duration=1000,animation}:IAnimateProps) => {

    const child = Children.only(children)
    const cssStyleDelay = addDelayCSSAnimation({elementProps:child.props,delay,duration})
    const newClassNames = addNewClassNamesAnimation({animation:animation,elementProps:child.props})
    

    const childProps:(Partial<any> & React.Attributes) = { ...child.props, 
        className: `animate ${newClassNames} `,
        style:cssStyleDelay,
        "data-testid":"animate-element"
    }

    return  cloneElement(children,childProps)
}
