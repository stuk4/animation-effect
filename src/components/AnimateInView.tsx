import { Children, cloneElement,useEffect } from 'react';
import { useInView } from 'react-intersection-observer';
import { IAnimateProps } from '../interfaces/IAnimations';
import { addDelayCSSAnimation, addNewClassNamesAnimation } from '../utils/animations';


interface AnimateInViewProps extends IAnimateProps{
    rootMargin?:string,
}

export const AnimateInView = ({
    children,
    duration=1000,
    delay=0,
    rootMargin='-100px 0px',
    animation}:AnimateInViewProps) => {


    const [ref, inView,node] = useInView({
        rootMargin,
        triggerOnce: true
    })

    const child = Children.only(children)
    const cssStyleDelay = addDelayCSSAnimation({elementProps:child.props,delay,duration})
    const newClassNames = addNewClassNamesAnimation({animation:animation,elementProps:child.props})

    const addAnimation = () => {
        const element = node?.target;
        if (!element) return;
        element.classList.add('animate');

    };
    useEffect(() => {
        if (!inView)return;
            addAnimation()
    }, [inView,addAnimation])
    
    const childProps:(Partial<any> & React.Attributes) = { ...child.props, 
        className: newClassNames,
        style:cssStyleDelay,
        ref
    }
  return  cloneElement(children,childProps)
}
