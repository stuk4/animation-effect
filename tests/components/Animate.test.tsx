import React from "react";
import { render, screen } from '@testing-library/react';
import { Animate } from '../../src/components/Animate';



const ReactTestRenderer = require('react-test-renderer');


describe('Animate component', () => {
  
  it('renders the children with animation styles',async () => {
    const text = 'Animated Text';
    const {getByTestId,rerender} = render(
  
      <Animate animation={"fadeIn"}>
        <p>{text}</p>
      </Animate>
    );
    const element = await screen.getByTestId("animate-element");
    await screen.debug()
    expect(element.classList.contains(`animate__${'fadeIn'}`)).toBe(true)
  

  });

});

it('matches the snapshot', () => {
  const text = 'Animated Text';
  const tree = ReactTestRenderer.create(
    <Animate animation="fadeIn">
        <p>{text}</p>
    </Animate>
    )
    .toJSON();
expect(tree).toMatchSnapshot();
});
// describe('TypeWriter component', () => {
//   afterEach(() => {
//     jest.clearAllMocks();
//   });

//   it('renders the correct element type', () => {
//     const tag = "h1"
//     const { getByTestId, } = render(
//       <TypeWriter elementType={tag} text="Test text" />
//     );
//     const element = getByTestId('typewriter-element');
//     expect(element.tagName.toLowerCase()).toBe(tag);
//   });


//   it('calls handleTypeWriter with correct parameters', async () => {
//     const text = 'Test text';
//     const interval = 20;
//     const delay = 0;
//     const { getByTestId } = render(
//       <TypeWriter elementType="h1" text={text} interval={interval} delay={delay} />
//     );
//     // waiting delay 
//     const element = getByTestId('typewriter-element')
//     await waitFor(() => {
//       expect(handleTypeWriter).toHaveBeenCalledWith(element, text, interval)
//     })

//   });



//   it('has the correct class name', () => {
//     const className = 'test-class';
//     const { getByTestId } = render(
//       <TypeWriter elementType="h1" text="Test text" classNames={className} />
//     );
//     const element = getByTestId('typewriter-element');
//     expect(element.classList.contains(className)).toBe(true);
   
//   });

  
//   it('matches the snapshot', () => {
//     const tree = ReactTestRenderer.create(<TypeWriter elementType="h1" text="Test text" />)
//       .toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });
