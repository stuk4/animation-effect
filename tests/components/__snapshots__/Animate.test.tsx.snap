// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<p
  className="animate animate__fadeIn "
  data-testid="animate-element"
  style={
    {
      "animationDuration": "1000ms",
    }
  }
>
  Animated Text
</p>
`;
